Add 'Create shortcut' button to instance toolbar

Leo40Git opened the pull request:
Closes #210.

This PR also makes `FileSystem::createShortcut` function on Windows systems.

### TODO
- [x] Icon for "Create Shortcut" action
  - Missing in Legacy, Flat, and Flat (White)
- [ ] Shortcut for "Create Shortcut" action
- [x] Platform-specific behavior
  - [x] Windows
    - [x] Use instance icon for shortcuts
  - [x] Linux
    - [x] Testing!!
  - [x] macOSX
    - [x] Write shell script
    - [x] Testing!!

Leo40Git committed:
Add 'Create shortcut' button to instance toolbar

Implemented on Windows only rn!

Signed-off-by: ADudeCalledLeo <7997354+Leo40Git@users.noreply.github.com>

flowln commented on launcher/ui/MainWindow.cpp:
```diff
@@ -2037,6 +2049,80 @@ void MainWindow::on_actionKillInstance_triggered()
+#include "shlguid.h"
+#endif
+
+void MainWindow::on_actionCreateInstanceShortcut_triggered()
```
there's already some barebones code for shortcut creation in [`FileSystem.cpp:341`](https://github.com/PrismLauncher/PrismLauncher/blob/54c26b05faa73dd6c8eb21f6fa898d438a3e867a/launcher/FileSystem.cpp#L341). Maybe you could use / do this stuff there instead?

Leo40Git replied:
oh whoops
Nice to know there's a way to do it on Windows _without_ having to do COM stuff, though apparently this path is broken? Will look at it later.

Leo40Git replied:
...no, wait, I think it's doing the same thing as I'm doing, damn

Leo40Git committed:
Windows: implement FS::createShortcut

Signed-off-by: ADudeCalledLeo <7997354+Leo40Git@users.noreply.github.com>

Leo40Git committed:
[UNTESTED] Linux: add icons to shortcuts

Signed-off-by: ADudeCalledLeo <7997354+Leo40Git@users.noreply.github.com>

Leo40Git commented on launcher/FileSystem.cpp:
```diff
@@ -338,12 +339,12 @@
 
 // Cross-platform Shortcut creation
-bool createShortCut(QString location, QString dest, QStringList args, QString name, QString icon)
+bool createShortcut(QString destination, QString target, QStringList args, QString name, QString icon)
```
I'll document this mess tomorrow (most of it is just "set" calls lmao)

Leo40Git commented on launcher/ui/MainWindow.cpp:
```diff
@@ -2037,6 +2050,45 @@ void MainWindow::on_actionKillInstance_triggered()
+        // TODO
+        // need to convert icon to ICO format and save it somewhere...
+        iconPath = "";
+#elif defined(Q_OS_UNIX)
```
```suggestion
#else
```

Leo40Git committed:
Remove unnessecary "is defined" check

Signed-off-by: ADudeCalledLeo <7997354+Leo40Git@users.noreply.github.com>

Leo40Git committed:
Add missing fail check for CoInitialize

Add a few comments

Signed-off-by: ADudeCalledLeo <7997354+Leo40Git@users.noreply.github.com>

Leo40Git committed:
Windows: add instance icon to shortcut

Signed-off-by: ADudeCalledLeo <7997354+Leo40Git@users.noreply.github.com>

Leo40Git commented:
Windows now has instance icons!

Leo40Git commented:
KNOWN BUG: Creating an instance shortcut on Windows replaces the application's icon with the instance's icon.
I have no idea how this happens (I blame COM), and am reasonably sure a simple `setIcon` call will fix this?

**EDIT:** Fixed!

Leo40Git committed:
Windows: fix window icon being replaced

Dunno why it happens but the fix was literally 2 lines, so whatever I guess

Signed-off-by: ADudeCalledLeo <7997354+Leo40Git@users.noreply.github.com>

Scrumplex commented on launcher/ui/MainWindow.cpp:
```diff
@@ -2050,6 +2063,73 @@ void MainWindow::on_actionKillInstance_triggered()
+            return;
+        }
+
+#if defined(Q_OS_MACOS)
+        // macOSX
+        // TODO actually write this path
+        QMessageBox::critical(this, tr("Create instance shortcut"), tr("Not supported on macOSX yet!"));
+#else
+        auto icon = APPLICATION->icons()->icon(m_selectedInstance->iconKey());
+
+        QString iconPath;
+        bool iconGenerated = false;
+
+#if defined(Q_OS_WIN)
+        iconPath = FS::PathCombine(m_selectedInstance->instanceRoot(), "icon.ico");
+
```
Having nested `#if` blocks might be confusing. Consider ending the macOS check with a return instead.

i.e.
```cpp
#ifdef Q_OS_MACOS
    // macOS
    // TODO actually write this path
    QMessageBox::critical(this, tr("Create instance shortcut"), tr("Creating shortcuts is not supported on macOS yet."));
    return;
#endif
```

Leo40Git committed:
Fix nested #if directive

Signed-off-by: ADudeCalledLeo <7997354+Leo40Git@users.noreply.github.com>

Scrumplex commented on launcher/FileSystem.cpp:
```diff
@@ -350,12 +351,12 @@ QString getDesktopDir()
-bool createShortCut(QString location, QString dest, QStringList args, QString name, QString icon)
+bool createShortcut(QString destination, QString target, QStringList args, QString name, QString icon)
 {
 #if defined(Q_OS_LINUX) || defined(Q_OS_FREEBSD)
```
```suggestion
#if defined(Q_OS_LINUX) || defined(Q_OS_FREEBSD) || defined(Q_OS_OPENBSD)
```

Scrumplex commented on launcher/ui/MainWindow.cpp:
```diff
@@ -776,6 +787,8 @@ class MainWindow::Ui
             }
         }
 
+        instanceToolBar->addAction(actionCreateInstanceShortcut); // TODO find better position for this
+
```
move this above the for loop, to make it left-aligned

Scrumplex commented on launcher/ui/MainWindow.cpp:
```diff
@@ -2047,6 +2060,73 @@ void MainWindow::on_actionKillInstance_triggered()
+            return;
+        }
+
+#if defined(Q_OS_MACOS)
+        // macOSX
+        // TODO actually write this path
+        QMessageBox::critical(this, tr("Create instance shortcut"), tr("Not supported on macOSX yet!"));
+		return;
+#endif
```
```suggestion
#ifdef Q_OS_MACOS
        QMessageBox::critical(this, tr("Create instance shortcut"), tr("Not supported on macOS yet!"));
        return;
#endif
```

Scrumplex commented on launcher/ui/MainWindow.cpp:
```diff
@@ -2047,6 +2060,73 @@ void MainWindow::on_actionKillInstance_triggered()
+        QString iconPath;
+        bool iconGenerated = false;
+
+#if defined(Q_OS_WIN)
```
```suggestion
#ifdef Q_OS_WIN
```

Scrumplex commented on launcher/ui/MainWindow.cpp:
```diff
@@ -2047,6 +2060,73 @@ void MainWindow::on_actionKillInstance_triggered()
+
+        // part of fix for weird bug involving the window icon being replaced
+        // dunno why it happens, but this 2-line fix seems to be enough, so w/e
+        auto appIcon = QGuiApplication::windowIcon();
```
```suggestion
        auto appIcon = Application::getThemedIcon("logo");
```

Leo40Git committed:
Apply suggestions from code review

Co-authored-by: Sefa Eyeoglu <contact@scrumplex.net>
Signed-off-by: ADudeCalledLeo <7997354+Leo40Git@users.noreply.github.com>

Leo40Git committed:
this ain't static, falco

Signed-off-by: ADudeCalledLeo <7997354+Leo40Git@users.noreply.github.com>

Leo40Git committed:
Move addAction call above left-align loop

Signed-off-by: ADudeCalledLeo <7997354+Leo40Git@users.noreply.github.com>

flowln commented on launcher/FileSystem.cpp:
```diff
@@ -350,12 +351,12 @@ QString getDesktopDir()
-#if defined(Q_OS_LINUX) || defined(Q_OS_FREEBSD)
-    location = PathCombine(location, name + ".desktop");
+#if defined(Q_OS_LINUX) || defined(Q_OS_FREEBSD) || defined(Q_OS_OPENBSD)
+    destination = PathCombine(destination, name + ".desktop");
```
```suggestion
    destination += ".desktop";
```

Leo40Git replied:
`destination` is the folder the shortcut will end up in. This will cause the `.desktop` file to be `/home/user/Desktop.desktop` (or wherever Qt thinks the desktop is).

Leo40Git replied:
nvm, disregard my earlier comment

flowln reviewed (commented):
please fix DCO :)

Leo40Git committed:
Linux: fix path shortcut is written to

Co-authored-by: flow <flowlnlnln@gmail.com>
Signed-off-by: ADudeCalledLeo <7997354+Leo40Git@users.noreply.github.com>

Leo40Git committed:
Mac: now supported! [UNTESTED]

Signed-off-by: ADudeCalledLeo <7997354+Leo40Git@users.noreply.github.com>

Leo40Git committed:
refactor #if checks

Signed-off-by: ADudeCalledLeo <7997354+Leo40Git@users.noreply.github.com>

Leo40Git committed:
Windows: remove icon if shortcut creation fails

Signed-off-by: ADudeCalledLeo <7997354+Leo40Git@users.noreply.github.com>

Leo40Git committed:
Mac: fix build

Signed-off-by: ADudeCalledLeo <7997354+Leo40Git@users.noreply.github.com>

Leo40Git committed:
Mac: attempt 2

- create .command files instead of .sh
- fix shortcuts not working if path to Prism Launcher contains spaces
- fix path to executable in shortcutss
- add check for running from extracted folder (prevents creating shortcuts)

Signed-off-by: ADudeCalledLeo <7997354+Leo40Git@users.noreply.github.com>

Leo40Git committed:
tabs -> spaces

Signed-off-by: ADudeCalledLeo <7997354+Leo40Git@users.noreply.github.com>

Leo40Git committed:
Linux: fixes

- Fix shortcut icons
- Possibly fix shortcut creation on AppImages
- Fix shortcut not working if path to launcher contains spaces

Signed-off-by: ADudeCalledLeo <7997354+Leo40Git@users.noreply.github.com>

Leo40Git committed:
Undo accidental tomlplusplus downgrade

Signed-off-by: ADudeCalledLeo <7997354+Leo40Git@users.noreply.github.com>

Leo40Git committed:
Linux: remove TryExec entry from .desktop files

Signed-off-by: ADudeCalledLeo <7997354+Leo40Git@users.noreply.github.com>

Leo40Git committed:
Windows, Linux: prevent segfault on missing icon

Signed-off-by: ADudeCalledLeo <7997354+Leo40Git@users.noreply.github.com>

flowln commented on launcher/ui/MainWindow.cpp:
```diff
@@ -735,6 +738,14 @@ class MainWindow::Ui
+        actionCreateInstanceShortcut.setTextId(QT_TRANSLATE_NOOP("MainWindow", "Create Shortcut"));
+        actionCreateInstanceShortcut.setTooltipId(QT_TRANSLATE_NOOP("MainWindow", "Creates a shortcut on your desktop to launch the selected instance."));
+        //actionCreateInstanceShortcut->setShortcut(QKeySequence(tr("Ctrl+D")));     // TODO
+        //actionCreateInstanceShortcut->setIcon(APPLICATION->getThemedIcon("copy")); // TODO
```
.

Leo40Git replied:
Icons are now implemented!
...well, kind of. The icon a duplicate of `externaltools` in most themes; and is notably missing in Legacy, Flat and Flat (White) - Flat and Flat (White)'s `externaltools` (a wrench) doesn't fit, and Legacy has a jigsaw piece which I can't even find in the files...

flowln reviewed (approved):
lgtm, thanks!

just do the TODOs (particularly the icon one, since now that we have icons on the sidebar for every other action, this action not having one looks weird) :D

Leo40Git committed:
Add icon

NOTE: Currently missing on Legacy, Flat and Flat (White)
Signed-off-by: ADudeCalledLeo <7997354+Leo40Git@users.noreply.github.com>

Scrumplex commented on launcher/resources/OSX/OSX.qrc:
```diff
@@ -1,5 +1,4 @@
-<!DOCTYPE RCC>
-<RCC version="1.0">
+<RCC>
```
what's up with this

(same for all files)

Leo40Git replied:
blame Qt Creator's built-in QRC editor

Leo40Git committed:
Restore QRC headers

Signed-off-by: ADudeCalledLeo <7997354+Leo40Git@users.noreply.github.com>

Leo40Git committed:
Merge remote-tracking branch 'upstream/develop' into feature/instance-shortcuts

Scrumplex reviewed (approved):
Thanks!

Scrumplex commented:
We can handle the missing icons in a separate PR
